internal/uvm/scsi/attach.go

```diff
@@ -0,0 +1,161 @@
+		// remove call for this slot, but is done for added safety.
+		slots[reservedSlot.Controller][reservedSlot.LUN] = &attachment{refCount: 1}
+	}
+	return &attachManager{
+		attacher:             attacher,
+		unplugger:            unplugger,
+		numControllers:       numControllers,
+		numLUNsPerController: numLUNsPerController,
+		slots:                slots,
+	}
+}
+
+type attachment struct {
+	controller uint
+	lun        uint
+	config     *AttachConfig
+	waitErr    error
+	waitCh     chan struct{}
+	refCount   uint
+}
+
+type AttachConfig struct {
+	Path     string
+	ReadOnly bool
+	Type     AttachmentType
+	EVDType  string // Only valid if Type is AttachmentTypeExtensibleVirtualDisk.
+}
+
+type AttachmentType string
+
+const (
+	AttachmentTypeVirtualDisk           AttachmentType = "VirtualDisk"
+	AttachmentTypePassThru              AttachmentType = "PassThru"
+	AttachmentTypeExtensibleVirtualDisk AttachmentType = "ExtensibleVirtualDisk"
+)
+
+func (am *attachManager) attach(ctx context.Context, c *AttachConfig) (controller uint, lun uint, err error) {
+	att, existed, err := am.trackAttachment(c)
+	if err != nil {
+		return 0, 0, err
+	}
+	if existed {
+		<-att.waitCh
+		if att.waitErr != nil {
+			return 0, 0, att.waitErr
+		}
+		return att.controller, att.lun, nil
+	}
+
+	defer func() {
+		if err != nil {
+			am.m.Lock()
+			am.untrackAttachment(att)
+			am.m.Unlock()
+		}
+
+		att.waitErr = err
+		close(att.waitCh)
+	}()
+
+	if err := am.attacher.attach(ctx, att.controller, att.lun, att.config); err != nil {
+		return 0, 0, fmt.Errorf("attach %s/%s at controller %d lun %d: %w", att.config.Type, att.config.Path, att.controller, att.lun, err)
+	}
+	return att.controller, att.lun, nil
+}
+
+func (am *attachManager) detach(ctx context.Context, controller, lun uint) (bool, error) {
+	am.m.Lock()
+	defer am.m.Unlock()
+
+	if controller >= uint(am.numControllers) || lun >= uint(am.numLUNsPerController) {
+		return false, fmt.Errorf("controller %d lun %d out of range", controller, lun)
+	}
+
+	att := am.slots[controller][lun]
+	att.refCount--
+	if att.refCount > 0 {
+		return false, nil
+	}
+
+	if err := am.unplugger.unplug(ctx, controller, lun); err != nil {
+		return false, fmt.Errorf("unplug controller %d lun %d: %w", controller, lun, err)
+	}
+	if err := am.attacher.detach(ctx, controller, lun); err != nil {
+		return false, fmt.Errorf("detach controller %d lun %d: %w", controller, lun, err)
+	}
+
+	am.untrackAttachment(att)
+
+	return true, nil
+}
+
+func (am *attachManager) trackAttachment(c *AttachConfig) (*attachment, bool, error) {
+	am.m.Lock()
+	defer am.m.Unlock()
+
+	var (
+		freeController int = -1
+		freeLUN        int = -1
+	)
+	for controller := range am.slots {
+		for lun := range am.slots[controller] {
+			attachment := am.slots[controller][lun]
+			if attachment == nil {
+				if freeController == -1 {
+					freeController = controller
+					freeLUN = lun
+				}
+			} else if c == attachment.config {
```

Comment:
This should be `*c == *attachment.config` or a call to `DeepEqual`

Reply:
Like the suggestion of `DeepEqual`, it appears to have the semantics we want here. Switched to it.